{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Tests for RNAfold</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PDB_ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>2D</th>\n",
       "      <th>Len</th>\n",
       "      <th>Structure</th>\n",
       "      <th>RNAfold_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1JVE</td>\n",
       "      <td>cctaattataacgaagttataattagg</td>\n",
       "      <td>(((((((((((((.)))))))))))))</td>\n",
       "      <td>27</td>\n",
       "      <td>Triplex-DNA</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1NGU</td>\n",
       "      <td>ctctccttgtatttcttacaaaaagag</td>\n",
       "      <td>(((((((((((.....)))))))))))</td>\n",
       "      <td>27</td>\n",
       "      <td>Hairpin with pseudoknots</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2OEY</td>\n",
       "      <td>ccatcgtctacctttggtaggatgg</td>\n",
       "      <td>(((((..(((((...))))))))))</td>\n",
       "      <td>25</td>\n",
       "      <td>Hairpin with loops</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2VWJ</td>\n",
       "      <td>aauggagacacggcttttgccgtgtc</td>\n",
       "      <td>......((((((((....))))))))</td>\n",
       "      <td>26</td>\n",
       "      <td>Hairpin with dangling ends</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3THW</td>\n",
       "      <td>cctctatctgaagccgatcgatgaagcatcgatcgcacagcttcag...</td>\n",
       "      <td>((((((((((((((((((((((((.))))))))))....)))))))...</td>\n",
       "      <td>53</td>\n",
       "      <td>Hairpin with loops</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PDB_ID                                           Sequence  \\\n",
       "0   1JVE                        cctaattataacgaagttataattagg   \n",
       "1   1NGU                        ctctccttgtatttcttacaaaaagag   \n",
       "2   2OEY                          ccatcgtctacctttggtaggatgg   \n",
       "3   2VWJ                         aauggagacacggcttttgccgtgtc   \n",
       "4   3THW  cctctatctgaagccgatcgatgaagcatcgatcgcacagcttcag...   \n",
       "\n",
       "                                                  2D  Len  \\\n",
       "0                        (((((((((((((.)))))))))))))   27   \n",
       "1                        (((((((((((.....)))))))))))   27   \n",
       "2                          (((((..(((((...))))))))))   25   \n",
       "3                         ......((((((((....))))))))   26   \n",
       "4  ((((((((((((((((((((((((.))))))))))....)))))))...   53   \n",
       "\n",
       "                    Structure  RNAfold_ref  \n",
       "0                 Triplex-DNA          1.0  \n",
       "1    Hairpin with pseudoknots          1.0  \n",
       "2          Hairpin with loops          1.0  \n",
       "3  Hairpin with dangling ends          1.0  \n",
       "4          Hairpin with loops          1.0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aptabase = pd.read_csv('Aptamers_2D.csv')\n",
    "aptabase.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(aptabase)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>RNAfold</h3>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;34mProcessing 1. input file \"test.fa\"\u001b[0m\n",
      "\u001b[33m>1JVE\u001b[0m\n",
      "ccuaauuauaacgaaguuauaauuagg\n",
      "((((((((((((...))))))))))))\u001b[32m (-11.00)\u001b[0m\n",
      "\u001b[33m>1NGU\u001b[0m\n",
      "cucuccuuguauuucuuacaaaaagag\n",
      "((((..(((((.....)))))..))))\u001b[32m ( -6.40)\u001b[0m\n",
      "\u001b[33m>230D\u001b[0m\n",
      "gggguuuugggguuuugggguuuugggg\n",
      "++++....++++....++++....++++\u001b[32m (-26.37)\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!RNAfold -g -v test.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: RNAfold [OPTIONS] [<input0.fa>] [<input1.fa>]...\n",
      "Calculate minimum free energy secondary structures and partition function of\n",
      "RNAs\n",
      "\n",
      "The program reads RNA sequences, calculates their minimum free energy (mfe)\n",
      "structure and prints the mfe structure in bracket notation and its free energy.\n",
      "If not specified differently using commandline arguments, input is accepted\n",
      "from stdin or read from an input file, and output printed to stdout. If the -p\n",
      "option was given it also computes the partition function (pf) and base pairing\n",
      "probability matrix, and prints the free energy of the thermodynamic ensemble,\n",
      "the frequency of the mfe structure in the ensemble, and the ensemble diversity\n",
      "to stdout.\n",
      "\n",
      "It also produces PostScript files with plots of the resulting secondary\n",
      "structure graph and a \"dot plot\" of the base pairing matrix.\n",
      "The dot plot shows a matrix of squares with area proportional to the pairing\n",
      "probability in the upper right half, and one square for each pair in the\n",
      "minimum free energy structure in the lower left half. For each pair i-j with\n",
      "probability p>10E-6 there is a line of the form\n",
      "\n",
      "i  j  sqrt(p)  ubox\n",
      "\n",
      "in the PostScript file, so that the pair probabilities can be easily extracted.\n",
      "\n",
      "Sequences may be provided in a simple text format where each sequence occupies\n",
      "a single line. Output files are named \"rna.ps\" and \"dot.ps\". Existing files\n",
      "of the same name will be overwritten.\n",
      "\n",
      "It is also possible to provide sequence data in FASTA format. In this case, the\n",
      "first word of the FASTA header will be used as prefix for output file names.\n",
      "PostScript files \"prefix_ss.ps\" and \"prefix_dp.ps\" are produced for the\n",
      "structure and dot plot, respectively. Note, however, that once FASTA input was\n",
      "provided all following sequences must be in FASTA format too.\n",
      "\n",
      "To avoid problems with certain operating systems and/or file systems the prefix\n",
      "will ALWAYS be sanitized! This step substitutes any special character in the\n",
      "prefix by a filename delimiter. See the --filename-delim option to change the\n",
      "delimiting character according to your requirements.\n",
      "\n",
      "The program will continue to read new sequences until a line consisting of the\n",
      "single character '@' or an end of file (EOF) condition is encountered.\n",
      "\n",
      "\n",
      "\n",
      "  -h, --help                    Print help and exit\n",
      "      --detailed-help           Print help, including all details and hidden\n",
      "                                  options, and exit\n",
      "      --full-help               Print help, including hidden options, and exit\n",
      "  -V, --version                 Print version and exit\n",
      "  -v, --verbose                 Be verbose.\n",
      "\n",
      "                                    (default=off)\n",
      "\n",
      "I/O Options:\n",
      "  Command line options for input and output (pre-)processing\n",
      "\n",
      "\n",
      "  -i, --infile=filename         Read a file instead of reading from stdin.\n",
      "\n",
      "  -o, --outfile[=filename]      Print output to file instead of stdout.\n",
      "\n",
      "  -j, --jobs[=number]           Split batch input into jobs and start\n",
      "                                  processing in parallel using multiple\n",
      "                                  threads. A value of 0 indicates to use as\n",
      "                                  many parallel threads as computation cores\n",
      "                                  are available.\n",
      "                                    (default=`0')\n",
      "      --noconv                  Do not automatically substitute nucleotide\n",
      "                                  \"T\" with \"U\".\n",
      "\n",
      "                                    (default=off)\n",
      "      --auto-id                 Automatically generate an ID for each sequence.\n",
      "                                    (default=off)\n",
      "      --id-prefix=STRING        Prefix for automatically generated IDs (as used\n",
      "                                  in output file names).\n",
      "                                    (default=`sequence')\n",
      "\n",
      "Algorithms:\n",
      "  Select additional algorithms which should be included in the calculations.\n",
      "  The Minimum free energy (MFE) and a structure representative are calculated\n",
      "  in any case.\n",
      "\n",
      "\n",
      "  -p, --partfunc[=INT]          Calculate the partition function and base\n",
      "                                  pairing probability matrix.\n",
      "                                    (default=`1')\n",
      "      --MEA[=gamma]             Compute MEA (maximum expected accuracy)\n",
      "                                  structure.\n",
      "                                    (default=`1.')\n",
      "  -c, --circ                    Assume a circular (instead of linear) RNA\n",
      "                                  molecule.\n",
      "\n",
      "                                    (default=off)\n",
      "  -g, --gquad                   Incoorporate G-Quadruplex formation into the\n",
      "                                  structure prediction algorithm.\n",
      "\n",
      "                                    (default=off)\n",
      "\n",
      "Structure Constraints:\n",
      "  Command line options to interact with the structure constraints feature of\n",
      "  this program\n",
      "\n",
      "\n",
      "      --maxBPspan=INT           Set the maximum base pair span.\n",
      "\n",
      "                                    (default=`-1')\n",
      "  -C, --constraint[=filename]   Calculate structures subject to constraints.\n",
      "                                    (default=`')\n",
      "      --enforceConstraint       Enforce base pairs given by round brackets '('\n",
      "                                  ')' in structure constraint.\n",
      "\n",
      "                                    (default=off)\n",
      "      --shape=filename          Use SHAPE reactivity data to guide structure\n",
      "                                  predictions.\n",
      "\n",
      "\n",
      "      --shapeConversion=method  Select method for SHAPE reactivity conversion.\n",
      "\n",
      "                                    (default=`O')\n",
      "\n",
      "Energy Parameters:\n",
      "  Energy parameter sets can be adapted or loaded from user-provided input files\n",
      "\n",
      "\n",
      "  -T, --temp=DOUBLE             Rescale energy parameters to a temperature of\n",
      "                                  temp C. Default is 37C.\n",
      "\n",
      "                                    (default=`37.0')\n",
      "  -P, --paramFile=paramfile     Read energy parameters from paramfile, instead\n",
      "                                  of using the default parameter set.\n",
      "\n",
      "      --salt=DOUBLE             Set salt concentration in molar (M). Default is\n",
      "                                  1.021M.\n",
      "\n",
      "\n",
      "  -m, --modifications[=STRING]  Allow for modified bases within the RNA\n",
      "                                  sequence string.\n",
      "                                    (default=`7I6P9D')\n",
      "      --mod-file=STRING         Use additional modified base data from JSON\n",
      "                                  file.\n",
      "\n",
      "\n",
      "\n",
      "Model Details:\n",
      "  Tweak the energy model and pairing rules additionally using the following\n",
      "  parameters\n",
      "\n",
      "\n",
      "  -d, --dangles=INT             How to treat \"dangling end\" energies for\n",
      "                                  bases adjacent to helices in free ends and\n",
      "                                  multi-loops.\n",
      "                                    (default=`2')\n",
      "      --noLP                    Produce structures without lonely pairs\n",
      "                                  (helices of length 1).\n",
      "                                    (default=off)\n",
      "\n",
      "Plotting:\n",
      "  Command line options for changing the default behavior of structure layout\n",
      "  and pairing probability plots\n",
      "\n",
      "\n",
      "      --noPS                    Do not produce postscript drawing of the mfe\n",
      "                                  structure.\n",
      "\n",
      "                                    (default=off)\n",
      "      --noDP                    Do not produce dot-plot postscript file\n",
      "                                  containing base pair or stack\n",
      "                                  probabilitities.\n",
      "                                    (default=off)\n",
      "\n",
      "If in doubt our program is right, nature is at fault.\n",
      "Comments should be sent to rna@tbi.univie.ac.at.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!RNAfold -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm *ps\n",
    "!rm *fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def aptamers_to_fasta(db, filename):\n",
    "\n",
    "    with open(filename, 'w') as fasta:\n",
    "\n",
    "        ids = list(db['PDB_ID'])\n",
    "        seqs = list(db['Sequence'])\n",
    "\n",
    "        for i in range(len(ids)):\n",
    "            print('>'+ids[i], file=fasta)\n",
    "            print(seqs[i], file=fasta)\n",
    "\n",
    "aptamers_to_fasta(aptabase, 'Aptamers_2D.fasta')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;34mProcessing 1. input file \"Aptamers_2D.fasta\"\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!RNAfold -g -v Aptamers_2D.fasta --outfile=rnafold_predictions.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def fastaparser(filename):\n",
    "\n",
    "    with open(filename, 'r') as fasta:\n",
    "        lines = fasta.readlines()\n",
    "        preds = []\n",
    "        for i in range(2, len(lines), 3):\n",
    "            pred = lines[i].split()\n",
    "            preds.append(pred[0])\n",
    "\n",
    "    return preds\n",
    "\n",
    "rnafold_preds = fastaparser('rnafold_predictions.fasta')\n",
    "len(rnafold_preds)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.96"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def accuracy_2D(pred, ref):\n",
    "\n",
    "    count = 0 \n",
    "    for k in range(len(pred)):\n",
    "\n",
    "        if pred[k] == ref[k]:\n",
    "            count += 1\n",
    "\n",
    "        elif (pred[k]=='+') and (ref[k] in ('(', ')', '[', ']', '{', '}')):\n",
    "            count += 1\n",
    "\n",
    "        elif (pred[k] in ('[', '{')) and (ref[k] in ('[', '{')):\n",
    "            count +=1\n",
    "\n",
    "        elif (pred[k] in (']', '}')) and (ref[k] in (']', '}')):\n",
    "            count +=1\n",
    "\n",
    "    return round(count/len(pred), 2)\n",
    "\n",
    "accuracy_2D(rnafold_preds[10], list(aptabase['2D'])[10])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PDB_ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>2D</th>\n",
       "      <th>Len</th>\n",
       "      <th>Structure</th>\n",
       "      <th>RNAfold_ref</th>\n",
       "      <th>RNAfold_2D</th>\n",
       "      <th>RNAfold_real</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1JVE</td>\n",
       "      <td>cctaattataacgaagttataattagg</td>\n",
       "      <td>(((((((((((((.)))))))))))))</td>\n",
       "      <td>27</td>\n",
       "      <td>Triplex-DNA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>((((((((((((...))))))))))))</td>\n",
       "      <td>0.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1NGU</td>\n",
       "      <td>ctctccttgtatttcttacaaaaagag</td>\n",
       "      <td>(((((((((((.....)))))))))))</td>\n",
       "      <td>27</td>\n",
       "      <td>Hairpin with pseudoknots</td>\n",
       "      <td>1.00</td>\n",
       "      <td>((((..(((((.....)))))..))))</td>\n",
       "      <td>0.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2OEY</td>\n",
       "      <td>ccatcgtctacctttggtaggatgg</td>\n",
       "      <td>(((((..(((((...))))))))))</td>\n",
       "      <td>25</td>\n",
       "      <td>Hairpin with loops</td>\n",
       "      <td>1.00</td>\n",
       "      <td>(((((..(((((...))))))))))</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2VWJ</td>\n",
       "      <td>aauggagacacggcttttgccgtgtc</td>\n",
       "      <td>......((((((((....))))))))</td>\n",
       "      <td>26</td>\n",
       "      <td>Hairpin with dangling ends</td>\n",
       "      <td>1.00</td>\n",
       "      <td>......((((((((....))))))))</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3THW</td>\n",
       "      <td>cctctatctgaagccgatcgatgaagcatcgatcgcacagcttcag...</td>\n",
       "      <td>((((((((((((((((((((((((.))))))))))....)))))))...</td>\n",
       "      <td>53</td>\n",
       "      <td>Hairpin with loops</td>\n",
       "      <td>1.00</td>\n",
       "      <td>(((((((((((((((((((((((...)))))))))....)))))))...</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4HT4</td>\n",
       "      <td>cgcgaacggaacgttcgcataagtgcgc</td>\n",
       "      <td>.((((((((.))))))))........()</td>\n",
       "      <td>28</td>\n",
       "      <td>Hairpin with dangling ends</td>\n",
       "      <td>1.00</td>\n",
       "      <td>.(((((((...)))))))..........</td>\n",
       "      <td>0.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1NGO</td>\n",
       "      <td>ctctttttgtaagaaatacaaggagag</td>\n",
       "      <td>(((((((((((.....)))))))))))</td>\n",
       "      <td>27</td>\n",
       "      <td>Hairpin</td>\n",
       "      <td>1.00</td>\n",
       "      <td>(((((((((((.....)))))))))))</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5N2Q</td>\n",
       "      <td>actttatgaaaataaagtatagtgtg</td>\n",
       "      <td>(((((((....)))))))........</td>\n",
       "      <td>26</td>\n",
       "      <td>Hairpin with dangling ends</td>\n",
       "      <td>1.00</td>\n",
       "      <td>(((((((....)))))))........</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3H25</td>\n",
       "      <td>cctttccccctacccgaagggtggggg</td>\n",
       "      <td>......(((((((((...)))))))))</td>\n",
       "      <td>27</td>\n",
       "      <td>Hairpin with dangling ends</td>\n",
       "      <td>1.00</td>\n",
       "      <td>......(((((((((...)))))))))</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1AW4</td>\n",
       "      <td>acctgggggagtattgcggaggaaggtaaacctgggggagtattgc...</td>\n",
       "      <td>(((((()([(((...))))..{)))))]}(((((()([(((...))...</td>\n",
       "      <td>27</td>\n",
       "      <td>Hairpin with pseudoknots</td>\n",
       "      <td>1.00</td>\n",
       "      <td>((((...(.((((((.(..(((........)))...).)))))).)...</td>\n",
       "      <td>0.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>6CCE</td>\n",
       "      <td>gcttgacaaaagtgttaaattgtgctatactagcacaatttaacac...</td>\n",
       "      <td>((((((((((((((((((((((((((.....)))))))))))))))...</td>\n",
       "      <td>57</td>\n",
       "      <td>Hairpin with dangling ends</td>\n",
       "      <td>0.96</td>\n",
       "      <td>(((((((((((((((((((((((((((...))))))))))))))))...</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1B4Y</td>\n",
       "      <td>tcttccttttccttctcccgagaaggtttt</td>\n",
       "      <td>.....(....(((((((..))))))))...</td>\n",
       "      <td>30</td>\n",
       "      <td>Triplex-DNA</td>\n",
       "      <td>0.60</td>\n",
       "      <td>....(((((((........)))))))....</td>\n",
       "      <td>0.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2N8A</td>\n",
       "      <td>gctggcttcgtaagaagccagctcgcggtcagcttgctgaccgcg</td>\n",
       "      <td>((((((((((..)))))))))).((((((((((..))))))))))</td>\n",
       "      <td>45</td>\n",
       "      <td>Double hairpin</td>\n",
       "      <td>0.95</td>\n",
       "      <td>(((((((((....))))))))).(((((((((....)))))))))</td>\n",
       "      <td>0.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3HXQ</td>\n",
       "      <td>ggcgtgcagtgccttcggccgtgcggtgcctccgtcacgcc</td>\n",
       "      <td>((((((.((.(((...))))).((((.(.).))))))))))</td>\n",
       "      <td>41</td>\n",
       "      <td>Two-forked with pseudoknots</td>\n",
       "      <td>0.65</td>\n",
       "      <td>((((.....))))...(((.((((((.....))).))))))</td>\n",
       "      <td>0.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>5HRT</td>\n",
       "      <td>cctggacggaaccagaatacttttggtctccagg</td>\n",
       "      <td>((((((((.)(((((((...))))))))))))))</td>\n",
       "      <td>34</td>\n",
       "      <td>Hairpin with loops</td>\n",
       "      <td>0.88</td>\n",
       "      <td>((((((..((.((((((...))))))))))))))</td>\n",
       "      <td>0.85</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PDB_ID                                           Sequence  \\\n",
       "0    1JVE                        cctaattataacgaagttataattagg   \n",
       "1    1NGU                        ctctccttgtatttcttacaaaaagag   \n",
       "2    2OEY                          ccatcgtctacctttggtaggatgg   \n",
       "3    2VWJ                         aauggagacacggcttttgccgtgtc   \n",
       "4    3THW  cctctatctgaagccgatcgatgaagcatcgatcgcacagcttcag...   \n",
       "5    4HT4                       cgcgaacggaacgttcgcataagtgcgc   \n",
       "6    1NGO                        ctctttttgtaagaaatacaaggagag   \n",
       "7    5N2Q                         actttatgaaaataaagtatagtgtg   \n",
       "8    3H25                        cctttccccctacccgaagggtggggg   \n",
       "9    1AW4  acctgggggagtattgcggaggaaggtaaacctgggggagtattgc...   \n",
       "10   6CCE  gcttgacaaaagtgttaaattgtgctatactagcacaatttaacac...   \n",
       "11   1B4Y                     tcttccttttccttctcccgagaaggtttt   \n",
       "12   2N8A      gctggcttcgtaagaagccagctcgcggtcagcttgctgaccgcg   \n",
       "13   3HXQ          ggcgtgcagtgccttcggccgtgcggtgcctccgtcacgcc   \n",
       "14   5HRT                 cctggacggaaccagaatacttttggtctccagg   \n",
       "\n",
       "                                                   2D  Len  \\\n",
       "0                         (((((((((((((.)))))))))))))   27   \n",
       "1                         (((((((((((.....)))))))))))   27   \n",
       "2                           (((((..(((((...))))))))))   25   \n",
       "3                          ......((((((((....))))))))   26   \n",
       "4   ((((((((((((((((((((((((.))))))))))....)))))))...   53   \n",
       "5                        .((((((((.))))))))........()   28   \n",
       "6                         (((((((((((.....)))))))))))   27   \n",
       "7                          (((((((....)))))))........   26   \n",
       "8                         ......(((((((((...)))))))))   27   \n",
       "9   (((((()([(((...))))..{)))))]}(((((()([(((...))...   27   \n",
       "10  ((((((((((((((((((((((((((.....)))))))))))))))...   57   \n",
       "11                     .....(....(((((((..))))))))...   30   \n",
       "12      ((((((((((..)))))))))).((((((((((..))))))))))   45   \n",
       "13          ((((((.((.(((...))))).((((.(.).))))))))))   41   \n",
       "14                 ((((((((.)(((((((...))))))))))))))   34   \n",
       "\n",
       "                      Structure  RNAfold_ref  \\\n",
       "0                   Triplex-DNA         1.00   \n",
       "1      Hairpin with pseudoknots         1.00   \n",
       "2            Hairpin with loops         1.00   \n",
       "3    Hairpin with dangling ends         1.00   \n",
       "4            Hairpin with loops         1.00   \n",
       "5    Hairpin with dangling ends         1.00   \n",
       "6                       Hairpin         1.00   \n",
       "7    Hairpin with dangling ends         1.00   \n",
       "8    Hairpin with dangling ends         1.00   \n",
       "9      Hairpin with pseudoknots         1.00   \n",
       "10   Hairpin with dangling ends         0.96   \n",
       "11                  Triplex-DNA         0.60   \n",
       "12               Double hairpin         0.95   \n",
       "13  Two-forked with pseudoknots         0.65   \n",
       "14           Hairpin with loops         0.88   \n",
       "\n",
       "                                           RNAfold_2D  RNAfold_real  \n",
       "0                         ((((((((((((...))))))))))))          0.93  \n",
       "1                         ((((..(((((.....)))))..))))          0.85  \n",
       "2                           (((((..(((((...))))))))))          1.00  \n",
       "3                          ......((((((((....))))))))          1.00  \n",
       "4   (((((((((((((((((((((((...)))))))))....)))))))...          0.96  \n",
       "5                        .(((((((...)))))))..........          0.86  \n",
       "6                         (((((((((((.....)))))))))))          1.00  \n",
       "7                          (((((((....)))))))........          1.00  \n",
       "8                         ......(((((((((...)))))))))          1.00  \n",
       "9   ((((...(.((((((.(..(((........)))...).)))))).)...          0.26  \n",
       "10  (((((((((((((((((((((((((((...))))))))))))))))...          0.96  \n",
       "11                     ....(((((((........)))))))....          0.60  \n",
       "12      (((((((((....))))))))).(((((((((....)))))))))          0.91  \n",
       "13          ((((.....))))...(((.((((((.....))).))))))          0.59  \n",
       "14                 ((((((..((.((((((...))))))))))))))          0.85  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aptabase.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnafold_accuracy_points = []\n",
    "ref2d = list(aptabase['2D'])\n",
    "\n",
    "for i in range(len(ref2d)):\n",
    "    rnafold_accuracy_points.append(accuracy_2D(rnafold_preds[i], ref2d[i]))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "% of Accurate structures (accuracy >= 0.85) = 49.275362318840585\n"
     ]
    }
   ],
   "source": [
    "#% of accurate structures\n",
    "acc_counter = 0\n",
    "for ac in rnafold_accuracy_points:\n",
    "    if ac >= 0.85:\n",
    "        acc_counter +=1\n",
    "\n",
    "print('% of Accurate structures (accuracy >= 0.85) =', 100*(acc_counter/len(rnafold_accuracy_points)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nucleotide Accuracy= 0.76\n"
     ]
    }
   ],
   "source": [
    "#Full accuracy\n",
    "nt_accuracy = accuracy_2D(''.join(rnafold_preds), ''.join(ref2d))\n",
    "print('Nucleotide Accuracy=', nt_accuracy)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we take into account non-canonical base pairs, % of accurate structures and average accuracy of RNAfold on dataset of short DNA aptamers is significantly lower than claimed in paper (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6975895/):\n",
    "\n",
    "Claimed % of accurate structures = 64%\n",
    "\n",
    "Real % of accurate structures = 49%\n",
    "\n",
    "Claimed nucleotide accuracy = 0.84\n",
    "\n",
    "Real nucleotide accuracy = 0.76"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PDB_ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>2D</th>\n",
       "      <th>Len</th>\n",
       "      <th>Structure</th>\n",
       "      <th>RNAfold_ref</th>\n",
       "      <th>RNAfold_2D</th>\n",
       "      <th>RNAfold_real</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1JVE</td>\n",
       "      <td>cctaattataacgaagttataattagg</td>\n",
       "      <td>(((((((((((((.)))))))))))))</td>\n",
       "      <td>27</td>\n",
       "      <td>Triplex-DNA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>((((((((((((...))))))))))))</td>\n",
       "      <td>0.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1NGU</td>\n",
       "      <td>ctctccttgtatttcttacaaaaagag</td>\n",
       "      <td>(((((((((((.....)))))))))))</td>\n",
       "      <td>27</td>\n",
       "      <td>Hairpin with pseudoknots</td>\n",
       "      <td>1.00</td>\n",
       "      <td>((((..(((((.....)))))..))))</td>\n",
       "      <td>0.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2OEY</td>\n",
       "      <td>ccatcgtctacctttggtaggatgg</td>\n",
       "      <td>(((((..(((((...))))))))))</td>\n",
       "      <td>25</td>\n",
       "      <td>Hairpin with loops</td>\n",
       "      <td>1.00</td>\n",
       "      <td>(((((..(((((...))))))))))</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2VWJ</td>\n",
       "      <td>aauggagacacggcttttgccgtgtc</td>\n",
       "      <td>......((((((((....))))))))</td>\n",
       "      <td>26</td>\n",
       "      <td>Hairpin with dangling ends</td>\n",
       "      <td>1.00</td>\n",
       "      <td>......((((((((....))))))))</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3THW</td>\n",
       "      <td>cctctatctgaagccgatcgatgaagcatcgatcgcacagcttcag...</td>\n",
       "      <td>((((((((((((((((((((((((.))))))))))....)))))))...</td>\n",
       "      <td>53</td>\n",
       "      <td>Hairpin with loops</td>\n",
       "      <td>1.00</td>\n",
       "      <td>(((((((((((((((((((((((...)))))))))....)))))))...</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>4U5M</td>\n",
       "      <td>tggtggtggtggttgtggtggtggtgtt</td>\n",
       "      <td>(([.)])([.)]..(.[).](.[).]..</td>\n",
       "      <td>28</td>\n",
       "      <td>G-quadruplex</td>\n",
       "      <td>0.71</td>\n",
       "      <td>.++.++.++.++................</td>\n",
       "      <td>0.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>2N3M</td>\n",
       "      <td>tggtggtggttgttgtggtggtggtggt</td>\n",
       "      <td>(([.)].[{.)]..}.([.)].([.)].</td>\n",
       "      <td>28</td>\n",
       "      <td>G-quadruplex</td>\n",
       "      <td>0.71</td>\n",
       "      <td>................++.++.++.++.</td>\n",
       "      <td>0.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>201D</td>\n",
       "      <td>ggggttttggggttttggggttttgggg</td>\n",
       "      <td>(((((.)[))))....((((...]))))</td>\n",
       "      <td>28</td>\n",
       "      <td>G-quadruplex</td>\n",
       "      <td>1.00</td>\n",
       "      <td>++++....++++....++++....++++</td>\n",
       "      <td>0.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>5J6U</td>\n",
       "      <td>ggggtttggggttttggggaagggg</td>\n",
       "      <td>((((.[.))))....(((]...)))</td>\n",
       "      <td>25</td>\n",
       "      <td>G-quadruplex</td>\n",
       "      <td>1.00</td>\n",
       "      <td>++++...++++....++++..++++</td>\n",
       "      <td>0.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>230D</td>\n",
       "      <td>ggggtutuggggttttgggguuttgggg</td>\n",
       "      <td>((((...)[)))(.).(((.....])))</td>\n",
       "      <td>28</td>\n",
       "      <td>G-quadruplex</td>\n",
       "      <td>1.00</td>\n",
       "      <td>++++....++++....++++....++++</td>\n",
       "      <td>0.86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   PDB_ID                                           Sequence  \\\n",
       "0    1JVE                        cctaattataacgaagttataattagg   \n",
       "1    1NGU                        ctctccttgtatttcttacaaaaagag   \n",
       "2    2OEY                          ccatcgtctacctttggtaggatgg   \n",
       "3    2VWJ                         aauggagacacggcttttgccgtgtc   \n",
       "4    3THW  cctctatctgaagccgatcgatgaagcatcgatcgcacagcttcag...   \n",
       "..    ...                                                ...   \n",
       "64   4U5M                       tggtggtggtggttgtggtggtggtgtt   \n",
       "65   2N3M                       tggtggtggttgttgtggtggtggtggt   \n",
       "66   201D                       ggggttttggggttttggggttttgggg   \n",
       "67   5J6U                          ggggtttggggttttggggaagggg   \n",
       "68   230D                       ggggtutuggggttttgggguuttgggg   \n",
       "\n",
       "                                                   2D  Len  \\\n",
       "0                         (((((((((((((.)))))))))))))   27   \n",
       "1                         (((((((((((.....)))))))))))   27   \n",
       "2                           (((((..(((((...))))))))))   25   \n",
       "3                          ......((((((((....))))))))   26   \n",
       "4   ((((((((((((((((((((((((.))))))))))....)))))))...   53   \n",
       "..                                                ...  ...   \n",
       "64                       (([.)])([.)]..(.[).](.[).]..   28   \n",
       "65                       (([.)].[{.)]..}.([.)].([.)].   28   \n",
       "66                       (((((.)[))))....((((...]))))   28   \n",
       "67                          ((((.[.))))....(((]...)))   25   \n",
       "68                       ((((...)[)))(.).(((.....])))   28   \n",
       "\n",
       "                     Structure  RNAfold_ref  \\\n",
       "0                  Triplex-DNA         1.00   \n",
       "1     Hairpin with pseudoknots         1.00   \n",
       "2           Hairpin with loops         1.00   \n",
       "3   Hairpin with dangling ends         1.00   \n",
       "4           Hairpin with loops         1.00   \n",
       "..                         ...          ...   \n",
       "64                G-quadruplex         0.71   \n",
       "65                G-quadruplex         0.71   \n",
       "66                G-quadruplex         1.00   \n",
       "67                G-quadruplex         1.00   \n",
       "68                G-quadruplex         1.00   \n",
       "\n",
       "                                           RNAfold_2D  RNAfold_real  \n",
       "0                         ((((((((((((...))))))))))))          0.93  \n",
       "1                         ((((..(((((.....)))))..))))          0.85  \n",
       "2                           (((((..(((((...))))))))))          1.00  \n",
       "3                          ......((((((((....))))))))          1.00  \n",
       "4   (((((((((((((((((((((((...)))))))))....)))))))...          0.96  \n",
       "..                                                ...           ...  \n",
       "64                       .++.++.++.++................          0.64  \n",
       "65                       ................++.++.++.++.          0.64  \n",
       "66                       ++++....++++....++++....++++          0.86  \n",
       "67                          ++++...++++....++++..++++          0.92  \n",
       "68                       ++++....++++....++++....++++          0.86  \n",
       "\n",
       "[69 rows x 8 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aptabase['RNAfold_2D'] = pd.Series(rnafold_preds)\n",
    "aptabase['RNAfold_real'] = pd.Series(rnafold_accuracy_points)\n",
    "aptabase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "aptabase.to_csv('RNAfold_benchmarking.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def TN(pred, ref):\n",
    "\n",
    "    counter = 0\n",
    "    for i in range(len(ref)):\n",
    "        if (pred[i] == '.') and (ref[i]=='.'):\n",
    "            counter +=1\n",
    "\n",
    "    return counter\n",
    "\n",
    "def FN(pred, ref):\n",
    "\n",
    "    counter = 0\n",
    "    for i in range(len(ref)):\n",
    "        if (pred[i] == '.') and (ref[i]!='.'):\n",
    "            counter +=1\n",
    "\n",
    "    return counter\n",
    "\n",
    "\n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_basepairs(s):\n",
    "    stack = []\n",
    "    #result = {}\n",
    "    result = []\n",
    "\n",
    "    for index, char in enumerate(s):\n",
    "        if char == '(':\n",
    "            # Запоминаем индекс открывающей скобки\n",
    "            stack.append(index)\n",
    "        elif char == ')':\n",
    "            # Если нашли закрывающую скобку, сопоставляем с последней открывающей\n",
    "            if stack:\n",
    "                open_index = stack.pop()\n",
    "                #result[open_index] = index\n",
    "                result.append((open_index, index))\n",
    "\n",
    "    return result\n",
    "\n",
    "def find_knotpairs(s):\n",
    "    stack = []\n",
    "    #result = {}\n",
    "    result = []\n",
    "\n",
    "    for index, char in enumerate(s):\n",
    "        if char == '[':\n",
    "            # Запоминаем индекс открывающей скобки\n",
    "            stack.append(index)\n",
    "        elif char == ']':\n",
    "            # Если нашли закрывающую скобку, сопоставляем с последней открывающей\n",
    "            if stack:\n",
    "                open_index = stack.pop()\n",
    "                #result[open_index] = index\n",
    "                result.append((open_index, index))\n",
    "\n",
    "    return result\n",
    "\n",
    "def find_otherpairs(s):\n",
    "    stack = []\n",
    "    #result = {}\n",
    "    result = []\n",
    "\n",
    "    for index, char in enumerate(s):\n",
    "        if char == '{':\n",
    "            # Запоминаем индекс открывающей скобки\n",
    "            stack.append(index)\n",
    "        elif char == '}':\n",
    "            # Если нашли закрывающую скобку, сопоставляем с последней открывающей\n",
    "            if stack:\n",
    "                open_index = stack.pop()\n",
    "                #result[open_index] = index\n",
    "                result.append((open_index, index))\n",
    "\n",
    "    return result\n",
    "\n",
    "def TP(pred,ref):\n",
    "\n",
    "    counter=0\n",
    "\n",
    "    pred_basepairs = find_basepairs(pred)\n",
    "    pred_knotpairs = find_knotpairs(pred)\n",
    "    pred_otherpairs = find_otherpairs(pred)\n",
    "\n",
    "    ref_basepairs = find_basepairs(ref)\n",
    "    ref_knotpairs = find_knotpairs(ref)\n",
    "    ref_otherpairs = find_otherpairs(ref)\n",
    "\n",
    "    for b in pred_basepairs:\n",
    "        if b in ref_basepairs:\n",
    "            counter += 2\n",
    "\n",
    "    for k in pred_knotpairs:\n",
    "        if k in ref_knotpairs:\n",
    "            counter += 2\n",
    "\n",
    "    for o in pred_otherpairs:\n",
    "        if k in ref_otherpairs:\n",
    "            counter += 2\n",
    "\n",
    "    return counter\n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Precision = 0.67\n",
      "Recall = 0.7\n",
      "F1-score = 0.68\n"
     ]
    }
   ],
   "source": [
    "tp = TP(''.join(rnafold_preds), ''.join(ref2d))\n",
    "fn = FN(''.join(rnafold_preds), ''.join(ref2d))\n",
    "tn = TN(''.join(rnafold_preds), ''.join(ref2d))\n",
    "fp = len(''.join(rnafold_preds)) - (tp+fn+tn)\n",
    "\n",
    "Precision = tp/(tp+fp)\n",
    "Recall = tp/(tp+fn)\n",
    "F1 = 2*Precision*Recall/(Precision+Recall)\n",
    "\n",
    "print('Precision =', round(Precision, 2))\n",
    "print('Recall =', round(Recall, 2))\n",
    "print('F1-score =', round(F1, 2))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
